monkey/infection_monkey/post_breach/actions/communicate_as_new_user.py

```diff
@@ -31,88 +32,76 @@ def __init__(self):
+        import win32api
+
+        try:
+            with NewUser(username, PASSWORD) as new_user:
+                # Using os.path is OK, as this is on windows for sure
+                ping_app_path = os.path.join(os.environ["WINDIR"], "system32", "PING.exe")
+                if not os.path.exists(ping_app_path):
+                    PostBreachTelem(self, ("{} not found.".format(ping_app_path), False)).send()
+                    return  # Can't continue without ping.
+
+                try:
+                    # Open process as that user:
+                    # https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessasusera
+                    commandline = "{} {} {} {}".format(ping_app_path, "google.com", "-n", "2")
+                    process_handle = win32process.CreateProcessAsUser(
+                        new_user.get_logon_handle(),  # A handle to the primary token that represents a user.
+                        None,  # The name of the module to be executed.
+                        commandline,  # The command line to be executed.
+                        None,  # Process attributes
+                        None,  # Thread attributes
+                        True,  # Should inherit handles
+                        win32con.NORMAL_PRIORITY_CLASS,  # The priority class and the creation of the process.
+                        None,  # An environment block for the new process. If this parameter is NULL, the new process
+                        # uses the environment of the calling process.
+                        None,  # CWD. If this parameter is NULL, the new process will have the same current drive and
+                        # directory as the calling process.
+                        win32process.STARTUPINFO()  # STARTUPINFO structure.
+                        # https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfoa
+                    )
+
+                    PostBreachTelem(self,
+                                    (CREATED_PROCESS_AS_USER_WINDOWS_FORMAT.format(commandline, username), True)).send()
+
+                    win32api.CloseHandle(process_handle[0])  # Process handle
+                    win32api.CloseHandle(process_handle[1])  # Thread handle
+
+                except Exception as e:
+                    # TODO: if failed on 1314, we can try to add elevate the rights of the current user with the
+                    #  "Replace a process level token" right, using Local Security Policy editing. Worked, but only
+                    #  after reboot. So:
+                    #  1. need to decide if worth it, and then
+                    #  2. need to find how to do this using python...
+                    PostBreachTelem(self, (
+                        "Failed to open process as user {}. Error: {}".format(username, str(e)), False)).send()
+
+                    # Nothing more we can do. Leak the process handle.
```

Comment:
Wrong comment. If process wasn't successfully created, the handle won't exist

Reply:
Removed. Comments are the devil anyway